### PR TITLE
allow users to hide/show email address (closes #3358)

### LINKED ISSUE #3358: Hide email addresses everywhere
At the moment registered users can see the email address of other users. Could you please add an option, either globally in app.ini or as a user setting, to *hide the email address completely from everybody* (included both registered users and visitors)? Thanks

### PR DESCRIPTION
**Pull request by bill-auger**
per issue #3358 - this PR allows users to hide/show email address by adding a checkbox to the form on the profile settings page (also to admin edit form) and an additional boolean to the User model - this is working locally but i have not yet written any tests - i could add them to this PR or as a separate PR

one caveat i should mention is that i was not able to get the translation to render - please advise

also im not sure if i should have upgraded any version numbers - please advise

modifications:
* add ShowEmail bool to User model
* add show/hide email checkbox on user profile settings form
* add show/hide email checkbox on admin edit user form
* add ShowEmail to user profile settings post handler
* add ShowEmail to admin edit user post handler
* show/hide emails on user profile and users directory per User.ShowEmail bool
* add en/US translation for 'show_email'

### CODE REVIEW

**Comment by Unknwon** [comment]
Thanks your PR!

But since default is show email, I guess you should just use `Hide my email`, etc.?

**Comment by bill-auger** [comment]
can i ask about the translation? im not sure why it doesnt take - oh i forgot i had screenshots ready

user-profile-settings
admin-edit-user
user-profile
explore-users

**Comment by Unknwon** [comment]
Run `make` and restart should take effect.

**Comment by ftrtyyuukjnbvd** [comment]
@Unknwon The default should be to **hide** the email. Then users can choose to show their email if they want to...

**Comment by bill-auger** [comment]
ok thanks - after running `make` the translation worked

**Comment by Unknwon** [comment]
>But since default is show email, I guess you should just use Hide my email, etc.?

**Comment by ftrtyyuukjnbvd** [comment]
@Unknwon right now the default is to show the email. But this is wrong, it should not be like this. Hopefully this patch would change this, and set the email **not visible** by default.

**Comment by bill-auger** [comment]
IMHO i really do think that downstreams should have email hidden by default because i think the majority of users would want their email hidden by default - just take a quick survey around github and count how many users show their on their profile and how many do not

ill change it if you want but i would first suggest adding a new config key to let the admin decide this for themselves

**Comment by rugk** [comment]
> adding a new config key to let the admin decide this for themselves

I think this would be a good idea.

**Comment by Unknwon** [comment]
>adding a new config key to let the admin decide this for themselves

+1, Though this only would work for new users.

**Comment by bill-auger** [comment]
im not agreeing to that CLA - that wasnt there yesterday - all of my commits were submitted before this so you can take them or leave them as is - i will not be contributing to this project while such a stipulation is in place

as for notabug - ironically they have recently been considering other platforms to migrate to - gitlab has already been rejected out of hand because of it's CLA so i think it is safe to assume that the community will concur that this makes the move away from gogs imperative

**Comment by Unknwon** [comment]
Sorry to hear that.

**Comment by rugk** [comment]
No, no, no... :weary:

@bill-auger Please read through https://github.com/gogits/gogs/issues/3384.

**Comment by rugk** [comment]
CA has been removed. @Unknwon please reopen this PR and @bill-auger please feel good again. :smiley:

**Comment by Unknwon** [comment]
OK, please solve the conflict.

**Comment by ftrtyyuukjnbvd** [comment]
This is comical. Is there a CLA or isn't there a CLA? First you introduce it, then remove, then reintroduce again but the text is not available

>There is no CLA to sign for gogits/gogs
> (Repository not found in Database)

The real question however is why did you introduce a CLA?

**Comment by hhenkel** [comment]
@ftrtyyuukjnbvd https://github.com/gogits/gogs/issues/3384

**Comment by rugk** [comment]
TLDR: There is not CLA currently. As for the reason why it was introduced I am also curious, but I am happy with the current state.

**Comment by eretuuhrthertyfyg** [comment]
@rugk The CLA is still in place.

**Comment by Unknwon** [comment]
It is really interesting that you're guys are not solving the conflict but continue talking about a bug from integration service. I would rather close this thread and resent the PR if original author wants to.

**Comment by rugk** [comment]
@Unknwon I think the issue currently is that it was not clear to @ftrtyyuukjnbvd and @eretuuhrthertyfyg whether the CLA was actuially in place or not.

And I think @bill-auger can decide by him/herself whether to send a new PR or not. He/She has not said anything about the CLA since two days (where the CLA was in place), so you cannot say "original author wants to" send a new PR.
Thus it is not necessary to close this PR.

**Comment by eretuuhrthertyfyg** [comment]
> It is really interesting that you guys are not solving the conflict but continue talking about a bug from integration service.

@Unknwon what you call "conflict" actually means "CLA", because signing the CLA was actually the only "conflict". I don't understand what "bug" you're talking about, although I think "bug" means that you want the email visible by default, whereas the patch hides it by default.

**Comment by rugk** [comment]
I think the "bug from integration service" is that it was claimed that the CLA is still required although it is not. This is all that caused confusion here.

**Comment by eretuuhrthertyfyg** [comment]
@rugk The CLA was still required, there was a big yellow button saying to sign the CLA. On the other hand, if the CLA is no longer required, why didn't he just merge the code in?
